styleguide/cpp.md

```diff
@@ -0,0 +1,873 @@
+    { ... }
+
+private:
+    const TTreeTraverserOptions Options_;
+    // ...arbitrarily complex private fields and methods.
+}
+
+ITreeTraverserPtr CreateTreeTraverser(const TTreeTraverserOptions& options)
+{
+    return New<TTreeTraverser>(options);
+}
+```
+
+Pimpl is currently considered a legacy idiom, but may be used sometimes. A notable case is when the CoW-semantics is involved, or the pointer implementation may be changed due to other reasons.
+
+<!------------------------------------------------------------------------------------>
+
+### Logging
+
+Log messages follow quite strict rules. This results in an extreme convenience for debugging and searching for relevant log messages whenever you get used to it.
+
+Logging is done only using the `YT_LOG_XXX` macros. 
+
+Log messages are usually in past simple, present continuous or do not contain a verb at all. Log messages are not completed English sentences, they do not finish with dot. In case when multiple "sentences" must be present within a single message, separate them with a semicolon.
+
+```cpp
+YT_LOG_DEBUG("Starting fetching columnar statistics");
+YT_LOG_DEBUG("Finished fetching columnar statistics");
+YT_LOG_DEBUG("Fetched current offset (Offset: %v, WallTime: %v)", currentOffset, timer.GetElapsedTime());
+YT_LOG_FATAL("Unknown write command (Command: %v)", command);
+YT_LOG_INFO("Concurrent cache head statistics (ElementCount: %v)", cacheHead->GetElementCount());
+YT_LOG_INFO("Replay encountered skip record; multiplexed suffix is ignored (ChunkId: %v, RecordId: %v, RecordCount: %v)", chunkId, recordId, recordCount);
+```
+
+In context of a caught exception or received error, it is usually logged as a first argument before the message.
+
+```cpp
+try {
+    EndpointHostName_ = FQDNHostName();
+} catch (const std::exception& ex) {
+    YT_LOG_ERROR(ex, "Failed to resolve local host name");
+    EndpointHostName_ = "localhost";
+}
+
+if (!commandResultOrError.IsOK()) {
+    YT_LOG_ERROR(commandResultOrError, "Failed to apply change replica mode command");
+    failed = true;
+}
+```
+
+If the logged message contains variable parts, they are always put in the brackets after the message in form of `Key: Value` pairs separated by commas. This is important for many reasons: readability, searchability, ability to parse the log messages automatically, ability to group log messages by the same key, etc. Finally, the logger may be configured to add some fixed tags to all log messages, and this logic also depends on the fixed form of the log message.
+
+```cpp
+// Ok.
+YT_LOG_DEBUG("Tables fetched (TableCount: %v)", tableCount);
+// Not ok.
+YT_LOG_DEBUG("Fetched %v tables", tableCount);
+YT_LOG_DEBUG("Tables fetched (tableCount: %v)", tableCount);
+```
+
+### Errors
+
+All error and exception text must be written as if they were dedicated to an end user. They must also be written in a gramatically correct English; typicall they also do not contain dots, having semicolons instead. 
```

Comment:
typicall -> typically

Reply:
Spell checker was something I forgot for sure.